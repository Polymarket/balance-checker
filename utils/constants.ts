import { ethers } from "ethers";

const MATIC_NODE: string = "wss://ws-matic-mainnet.chainstacklabs.com";
const provider = new ethers.providers.WebSocketProvider(MATIC_NODE);
const { abi } = require("./ConditionalTokens.json");
const CONTRACT_ADDRESS: string = "0x4D97DCd97eC945f40cF65F87097ACe5EA0476045";

export const contract = new ethers.Contract(CONTRACT_ADDRESS, abi, provider);
export const parentCollectionId: string =
    "0x0000000000000000000000000000000000000000000000000000000000000000";
export const collateralToken = "0x2791Bca1f2de4661ED88A30C99A7a9449Aa84174";

export type Market = {
    id: number;
    question: string;
    conditionId: string;
    slug: string;
    twitter_card_image: string;
    resolution_source: any;
    end_date: string;
    category: string;
    amm_type: any;
    liquidity: any;
    sponsor_name: any;
    sponsor_image: any;
    start_date: any;
    x_axis_value: unknown;
    y_axis_value: unknown;
    denomination_token: any;
    fee: unknown;
    image: any;
    icon: any;
    lower_bound: any;
    upper_bound: any;
    description: any;
    tags: string[];
    outcomes: string[];
    outcomePrices: string[];
    volume: string;
    active: boolean;
    market_type: string;
    format_type: any;
    lower_bound_date: any;
    upper_bound_date: any;
    closed: boolean;
    marketMakerAddress: string;
    created_at: any;
    updated_at: any;
    closed_time: any;
    wide_format: unknown;
    new: any;
    sent_discord: any;
    mailchimp_tag: any;
    use_cases: unknown[];
    seo: any;
};
