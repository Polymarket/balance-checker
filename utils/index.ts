import {
    Market,
    contract,
    parentCollectionId,
    collateralToken,
} from "./constants";

/** 
 * getPositionId
 * This function is called by clicking on an outcome. It gets the position Id for that outcome
 * @param {Market} market - the market object
 * @param {string} outcome - name of the outcome
 * @param {func} setPositionId - parent state setter for positionId
 * @param {func} setErrorMessage - parent state setter for errorMessage
 * */
export async function getPositionId({
    market,
    outcome,
    setPositionId,
    setErrorMessage,
}: {
    market: Market;
    outcome: string;
    setPositionId: React.Dispatch<React.SetStateAction<string>>;
    setErrorMessage: React.Dispatch<React.SetStateAction<string>>;
}): Promise<string> {
    try {
        const outcomeArray: number[] = new Array(market.outcomes.length).fill(
            0,
        );
        const outcomeIndex: number = market.outcomes.indexOf(outcome);
        outcomeArray.splice(outcomeIndex, 1, 1);
        const binary: string = outcomeArray.reverse().join("");
        const indexSet: number = parseInt(binary, 2);
        const collectionId: string = await contract
            .getCollectionId(parentCollectionId, market.conditionId, indexSet);
            
        const positionId: string = await contract
            .getPositionId(collateralToken, collectionId);
            setPositionId(positionId);
        return positionId
    } catch (error) {
        setErrorMessage(error.message);
        return error.message;
    }
}


/** 
 * getBalance
 * This function is called by clicking submit in the modal. It gets the balance of the provided address.
 * @param {string} address - should be customers Polymarket account address, not their wallet address
 * @param {string} positionId - the id for the position returned from getPositionId
 * @param {func} setBalance - component state setter for positionId
 * @param {func} setErrorMessage - parent state setter for errorMessage
 * */
export async function getBalance({
    address,
    positionId,
    setBalance,
    setErrorMessage,
}: {
    address: string;
    positionId: string;
    setBalance: React.Dispatch<React.SetStateAction<string | undefined>>;
    setErrorMessage: React.Dispatch<React.SetStateAction<string>>;
}): Promise<string> {
    try {
        const balance: string = await contract
            .balanceOf(address, positionId);
       setBalance((+balance / 1000000).toFixed(6));
       return balance;
    } catch (error) {
         setErrorMessage(error.message);
         return (error.message)
    }
}

/** 
 * searchMarkets
 * This function is called by typing in the search input. It searches market questions and descriptions for matches
 * @param {Array} object - the array of market objects fetched by getStaticProps from the api
 * @param {Array} properties - keys for the properties of the market objects
 * @param {string} query - the text coming from the search input
 * */

export function searchMarkets<data>(
    object: data,
    properties: Array<keyof data>,
    query: string,
): boolean {
    if (query === "") {
        return true;
    }

    return properties.some((property) => {
        const value = object[property];
        if (typeof value === "string" || typeof value === "number") {
            return value.toString().toLowerCase().includes(query.toLowerCase());
        }
        return false;
    });
}