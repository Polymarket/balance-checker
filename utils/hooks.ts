import { Market, parentCollectionId, collateralToken } from "./constants";
import { getIndexSet, getCollectionId, getPositionId } from ".";

export async function getModalProps({
    market,
    outcome,
}: {
    market: Market;
    outcome: string;
}): Promise<{ outcomeState: string; question: string; positionId: string }> {
    const outcomeState = outcome;
    const { question } = market;
    const { conditionId } = market;
    const indexSet = getIndexSet({ market, outcome });
    const collectionId = await getCollectionId({
        parentCollectionId,
        conditionId,
        indexSet,
    });
    const positionId = await getPositionId({ collateralToken, collectionId });

    return { outcomeState, question, positionId };
}
