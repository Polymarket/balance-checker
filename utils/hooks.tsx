
import { Market, parentCollectionId, collateralToken } from "./constants";
import { getIndexSet, getCollectionId, getPositionId } from ".";

export async function getModalProps({ market, outcome }: { market: Market; outcome: string }): Promise<{ outcomeName: string; question: string; positionId: string; }> {
   
    
    const outcomeName = outcome
    const question = market.question;
    const conditionId = market.conditionId;
    const indexSet = getIndexSet({market, outcome});
    const collectionId = await getCollectionId({parentCollectionId, conditionId, indexSet});
    const  positionId = await getPositionId({ collateralToken, collectionId });
   
   
   
    return { outcomeName, question, positionId }

}