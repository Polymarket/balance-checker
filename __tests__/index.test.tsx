 import * as React from 'react';
 import { mount, shallow} from 'enzyme';
import{ Market } from "../utils/constants";
const fetch = require('node-fetch');
import { Form, Button, Alert, Modal, Card } from "react-bootstrap";
import { getBalance } from "../utils";
import nock from 'nock';
import SearchInput from "../components/searchInput";
import BalanceChecker from "../pages/index";
import ActiveMarkets from '../components/activeMarkets';
import BalanceModal from '../components/modal';

describe('<SearchInput/>', () => {
  it('should render correctly', async function () {
    const onChangeSearchQuery = jest.fn()
    const props = {
      onChangeSearchQuery
  }
    const wrap = shallow(<SearchInput {...props}/>);
    expect(wrap.find("label").text()).toBe("Search Markets");
  });

});
describe('getBalance', () => {


  it('should get customer balance', async function () {

    const setBalance = jest.fn();
    const setErrorMessage = jest.fn();
    const address = "0x822276EB1df687f4faB5D29adA0d30590C510311";
    const positionId = "0xf270803dfab4f5aa7f9e9a4996ce3425f219102f35ac062275fe19842bfd46fe";

    const props = {
      address,
      positionId,
      setBalance,
      setErrorMessage,
     
  }
    
    let balance = getBalance(props)
    expect(balance).toBe("0");
  });

});

let data:Market[];
beforeAll(async (done) => {
  const res = await fetch(
    `https://strapi-matic.poly.market/markets?active=true&_sort=volume:desc`,
);
  data = await res.json();
 done();
});


  describe('<BalanceChecker/>', () => {
    it('should render correctly', async function () {
      const wrap = mount(<BalanceChecker data={data}/>);
      expect(wrap.find('h1').text()).toBe('Customer Balances');

     
    });
    
    
    
    it('should open the balance modal', async function () {
      const wrap = mount(<BalanceChecker data={data}/>);
      wrap.findWhere(node => node.key() === 'Elon Musk').simulate("click");
      expect(wrap.find(Modal.Title).text()).toBe('Customer Balance');
    
    
     
    })
});

describe('<ActiveMarkets/>', () => {
  it('should render correctly', async function () {
    const handleClick = jest.fn();
    const query = "";
    const props = {
      data,
      query,
      handleClick
  }
    const wrap = shallow(<ActiveMarkets {...props}/>);
    expect(wrap.exists(Card)).toEqual(true);;

   
  });

})
    


