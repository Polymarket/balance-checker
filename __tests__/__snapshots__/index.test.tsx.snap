// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BalanceChecker/> should render correctly 1`] = `[Function]`;

exports[`<BalanceModal/> should render correctly 1`] = `[Function]`;

exports[`<SearchInput/> should render correctly 1`] = `[Function]`;
